### PR TITLE
feat: watch for changed integration test files

### PR DESCRIPTION
**Pull request by boukeversteegh**
This was quite some work, but well worth it.

Added

    yarn watch

Watch for changes in `integration/*/*.{bin, proto}` and automatically run the correct scripts (`proto2bin`, `proto2pbjs` and `bin2ts`).

Automatically keeps `.ts` and `.pbjs` files up to date while working on `ts-proto`.

## Motivation

- Remembering which scripts to run when stayed difficult, and I still messed it up from time to time.
- Running the scripts on all test files is slow
- Running the script on a single file is fast, but you need to type commands by hand, or create new profiles in your IDE every time you are working with a different integration test
- Improve overall developer experience

## Features

- Only processes the changed files

## Missing features

- Does not regenerate ALL test files automatically when you change the plugin implementation (e.g. `main.ts`).  
  This would be extremely impractical as it takes at least a minute, and would run each time you type a letter

## Implementation

- Uses [chokidar](https://github.com/paulmillr/chokidar) to watch for file changes
  - This is the library used internally by `nodemon`
  - `nodemon` cannot pass the changed files as arguments to the configured `exec` script, so using `nodemon` would force us to regenerate _all_ files, each time one changed.
  - [npm-watch](https://www.npmjs.com/package/npm-watch) is a nice wrapper around `nodemon` but has the same limitation, so couldn't be used
- All generation shell scripts had to be adapted:
  - Accepting an input file: i.e. `codegen.sh value/value.bin` (this is passed by `yarn watch`)
  - Keep backwards compatibility:
    - Running without arguments generates files for all tests
    - Specifying a directory name to only generate files for that test
- [integration/pbjs.sh](integration/pbjs.sh)
  - Still "by hand", no abstraction was introduced
  - Comments replaced by if statements to allow for partial execution
  - If new integration tests are added (i.e. from another PR) without the if statement, the script will still work, but execute that item unconditionally

## Screenshot

Running `yarn watch` and updating `integration/simple-optionals/simple.proto`

image

### CODE REVIEW

**Review by stephenh — APPROVED** [review]
Wow, this is amazing! I'm surprised at how small & neat the `watch.ts` script is too; I'm definitely going to be cribbing this for another use cases I have on other/future projects. :-D

Thanks!

File: package.json [hunk 1]
```diff
@@ -57,6 +58,7 @@
   },
   "dependencies": {
     "@types/object-hash": "^1.3.0",
+    "chokidar": "^3.5.2",
```
[the remark this answers is not in the document]
Comment (boukeversteegh) [reply package.json]: Done 👍

**Comment by boukeversteegh** [comment]
> Wow, this is amazing! I'm surprised at how small & neat the `watch.ts` script is too; I'm definitely going to be cribbing this for another use cases I have on other/future projects. :-D
> 
> Thanks!

I'm glad to hear that! I see that this doesn't work on Ubuntu in WSL, so I'm adding an option `--polling` to fix that.

Please let me know if it works on Linux.

Then I see there's still an issue generating the pbjs files, so I broke something there. Will fix that.

**Comment by stephenh** [comment]
@boukeversteegh wow, worked the 1st time on linux. `docker-compose run` even auto-built the images on the 1st run, and then just ran them directly on subsequent saves. Looks great!

**Pull request merged**